src/components/TodoContext/TodosContext.tsx

```diff
@@ -0,0 +1,381 @@
+    todoId: number,
+  ) => void,
+  handleTitleUpdate: (
+    event: React.FormEvent<HTMLFormElement> | null,
+    todo: Todo,
+  ) => void,
+  setTodoInputStatus: (id: number, todo: Todo | null) => void,
+  closeErrorMessage: () => void,
+  changeAllComplet: (isCompleted: boolean) => void,
+  unsaveTitle: () => void,
+};
+
+// #endregion
+
+export const TodoContext = React.createContext<Context>({
+  todos: [],
+  newTodo: null,
+  newTodoField: null,
+  error: '',
+  modifiedTodosId: [0],
+  hidden: true,
+  todoWithFormId: 0,
+});
+
+export const TodoUpdateContext = React.createContext<UpdateContext>({
+  handleNewSubmit: () => {},
+  handleNewInput: () => {},
+  clearCompleted: () => {},
+  removeTodo: () => {},
+  handleChangeComplet: () => {},
+  handleChangeTitle: () => {},
+  handleTitleUpdate: () => {},
+  setTodoInputStatus: () => {},
+  closeErrorMessage: () => {},
+  changeAllComplet: () => {},
+  unsaveTitle: () => {},
+});
+
+export const TodosProvider: React.FC<Props> = ({ children }) => {
+  const user = useContext(AuthContext);
+  const newTodoField = useRef<HTMLInputElement>(null);
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [newTodo, setNewTodo] = useState<Todo | null>(null);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [modifiedTodosId, setModifiedTodosId] = useState<number[]>([0]);
+  const [hidden, setHidden] = useState(true);
+  const [error, setError] = useState('');
+  const [todoWithFormId, setTodoWithFormId] = useState(0);
+
+  // #region ---- DATA API ----
+  // download todos from server
+  async function loadTodos() {
+    if (user) {
+      try {
+        const todosFromApi = await getTodos(user?.id);
+
+        setTodos(todosFromApi);
+      } catch {
+        setHidden(false);
+        setError('Unable to download todos');
+        throw new Error('Unable to download todos');
+      }
+    }
+  }
+
+  // adding todo when getting title from user
+  async function sendNewTodo(data: Todo) {
+    if (tempTodo) {
+      setTodos(current => (
+        [...current, { ...tempTodo, title: tempTodo.title.trim() }]
+      ));
+    }
+
+    if (user && data) {
+      try {
+        const todoFromServer = await sendTodo(user?.id, data);
+
+        setTodos(current => current.map(todo => {
+          if (todo.id === 0) {
+            return { ...todo, id: todoFromServer.id };
+          }
+
+          return todo;
+        }));
+      } catch {
+        setHidden(false);
+        setTodos(current => current.slice(0, current.length - 1));
+        setError('Unable to add a todo');
+        throw new Error('Unable to add a todo');
+      }
+    }
+
+    setNewTodo(null);
+  }
+
+  // delete todos or todo
+  async function deleteTodos(todosId: number[]) {
+    if (user && todosId) {
+      try {
+        await Promise.all(todosId.map(todoId => deleteTodo(todoId)));
+
+        setTodos(currentTodos => (
+          currentTodos.filter(({ id }) => !todosId.includes(id))
+        ));
```

Comment:
it is better to make a new request to the server for new tasks (optional comment, you can leave it as it is)